### PR TITLE
Introduce a cache to identify visited types

### LINKED ISSUE #30681: `org.wso2.ballerinalang.compiler.semantics.analyzer.Types#getIntersection` has not been updated for cyclic types
**Description:**
$title

**Steps to reproduce:**
The following results in an SO error.

```ballerina
type Foo int|Foo[]|object {};

function testImmutableImportedCyclicUnionVariable() {
    Foo x = 1;
    anydata m = <anydata> x;
}
```

**Affected Versions:**
slbeta1-rc1

### LINKED ISSUE #31011: BAD SAD while type testing a Complex Union type
Observed this in SLBeta1. The code I used here is 

```
if stmt is anydata {
            
}
```

Here `stmt` is https://github.com/ballerina-platform/nballerina/blob/9f3819aa1e8a12ce47261836ef11a947ed587913/modules/front/ast.bal#L19

Log: [ballerina-internal.log](https://github.com/ballerina-platform/ballerina-lang/files/6602994/ballerina-internal.log)

### PR DESCRIPTION
**Pull request by KRVPerera**
This fixes a stack overflow in getIntersection method.
Fixes #30681
Fixes #31011 

Fixes #<Issue Number>

## Remarks

TODO: 
check 

#30142 jVM generation is not supported for type other error with type guard

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/lib-creator/src/main/java/org/ballerinalang/stdlib/utils/BuildLangLib.java [hunk 1]
```diff
@@ -28,32 +28,22 @@
 import io.ballerina.projects.PackageManifest;
```
Comment (MaryamZi) [inline misc/lib-creator/src/main/java/org/ballerinalang/stdlib/utils/BuildLangLib.java:28]: These changes are not related to/required for this PR, right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -3479,18 +3480,24 @@ BType getTypeForFiniteTypeValuesAssignableToType(BFiniteType finiteType, BType t
...
+     * @param visitedTypes cache to capture visited types
      * @return           a single type or a new union type if at least one member type of the union type is
      *                      assignable to targetType, else semanticError
      */
     BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType targetType, SymbolEnv env,
-                                                     IntersectionContext intersectionContext) {
+                                                     IntersectionContext intersectionContext,
+                                                     LinkedHashSet<BType> visitedTypes) {
         List<BType> intersection = new LinkedList<>();
+        if (!visitedTypes.add(unionType)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:3494]: Don't we need the type pair here? Is it always enough to add just the union type?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java [hunk 3]
```diff
@@ -860,7 +860,8 @@ private boolean isDisjoint(BUnionType t1, BUnionType t2, BParameterizedType para
 
     private boolean hasIntersection(BType t1, BType t2) {
         BType typeIntersection =
-                types.getTypeIntersection(Types.IntersectionContext.compilerInternalIntersectionContext(), t1, t2, env);
+                types.getTypeIntersection(Types.IntersectionContext.compilerInternalIntersectionContext(), t1, t2, env,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java:863]: Can we introduce overloaded methods and avoid the changes in most places?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal [hunk 4]
```diff
@@ -760,6 +760,19 @@ function testTypeNarrowingForIntersectingDirectUnion_1() returns boolean {
     return false;
 }
 
+type CyclicComplexUnion int|CyclicComplexUnion[]|object {};
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal:763]: Can we add negative tests also? Where there's no intersection.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 5]
```diff
@@ -4127,7 +4140,8 @@ public BType getTypeIntersection(IntersectionContext intersectionContext, BType
         }
     }
 
-    private BType getIntersection(IntersectionContext intersectionContext, BType lhsType, SymbolEnv env, BType type) {
+    private BType getIntersection(IntersectionContext intersectionContext, BType lhsType, SymbolEnv env, BType type,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4143]: The intersection computation still seems to have issues, although it does not crash.

```ballerina
type Foo int|Foo[]|object {};
type Bar int|Bar[]|string;

public function main() {
    Foo s = <int[]> [1, 2];
    float|Foo m = <Foo>s;
    if (m is Bar|string) {
        int f2 = m; // no compilation error, panics at runtime
        panic error("1");
    }
    panic error("2");
}
```

Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4143]: This is an existing issue right

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal (hunk 4 above)
Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal:763]: done

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal:763]: Can we include a negative compilation test too?

Comment (KRVPerera) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard.bal:763]: added to `type-guard-negative.bal`

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard-negative.bal [hunk 6]
```diff
@@ -222,3 +222,26 @@ function testMapIntersectionNegative() {
         map<int> x = m;
     }
 }
+
+type CyclicComplexUnion int|CyclicComplexUnion[]|object {};
+type floatUnion float|string;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/ifelse/type-guard-negative.bal:227]: Let's add another test where the target type is also a cyclic union. We can do this separately.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**